mozphab/jujutsu.py

```diff
@@ -0,0 +1,433 @@
+            ).rstrip()
+        except:
+            pass
+        if current_revision_is_empty not in ["false", "true"]:
+            raise Error(
+                "Failed to parse current revision for unknown reasons."
+            )  # TODO: casing, etc. consistent?
+        self.__change_id_at_start_was_empty = current_revision_is_empty == "true"
+
+        try:
+            revset = "@-" if self.__change_id_at_start_was_empty else "@"
+            self.__previous_change_ids = self.__cli_log(
+                revset=revset, template="change_id"
+            )
+        except Exception:
+            raise Error("Jujutsu failed to read the branch name for unknown reasons.")
+
+    def set_args(self, args: argparse.Namespace):
+        """Store moz-phab command line args and set the revset."""
+        super().set_args(args)
+
+        # TODO: Did we accidentally change the behavior of `--single` with no `start_rev` specified?
+        is_single = hasattr(self.args, "single") and self.args.single
+
+        # TODO: What if we try to submit something that's immutable?
+        start_rev = None
+        if (
+            hasattr(self.args, "start_rev")
+            and self.args.start_rev != environment.DEFAULT_START_REV
+        ):
+            start_rev = self.args.start_rev
+        else:
+            start_rev = "@-" if is_single else self.__get_first_mutable_change()
+        if not start_rev:
+            return None
+
+        end_rev = None
+        if is_single:
+            end_rev = start_rev
+        elif (
+            hasattr(self.args, "end_rev")
+            and self.args.end_rev != environment.DEFAULT_END_REV
+        ):
+            end_rev = self.args.end_rev
+        else:
+            end_rev = start_rev
+
+        self.revset = (start_rev, end_rev)
+
+    def commit_stack(self, single: bool = False) -> Optional[List[Commit]]:
+        """Collect all the info about commits."""
+        logger.debug(f"searching with start and end at {self.revset}")
+        if not self.revset:
+            # No commits found to submit
+            return None
+
+        boundary = '"--%s--\n"' % uuid.uuid4().hex
+        log = self.__cli_log(
+            # XXX: Some duplication with `self.__cli_log`. Dedupe?
+            template="".join(
+                [
+                    "separate('\n', ",
+                    "author.timestamp().utc().format('%+'), ",
+                    "author.name(), " "author.email(), ",
+                    "self.parents().map(|c| c.commit_id().short()).join(' '), ",
+                    "change_id.short(), ",
+                    "commit_id, ",
+                    "conflict, ",
+                    "description, ",
+                    boundary,
+                    ")",
+                ]
+            ),
+            revset="{}::{}".format(self.revset[0], self.revset[1]),
+            split=False,
+            strip=False,
+        )[: -len(boundary)]
+        changes = log.split(boundary)
```

Comment:
Currently, the stack gets submitted in the wrong order, so you should probably invert the list of commits at some point.

Reply:
Uhm, I ran it today and my stack definitely arrived in the wrong order on Phabricator 😅.

![Screenshot_20240918_174721](https://github.com/user-attachments/assets/8265d3f0-7468-475d-b5a7-9ca6c8191334)

I think the point where the Git commits are being reversed is here: https://github.com/erichdongubler-mozilla/review/blob/901957058371ed9c502095d19526ac0b94bec722/mozphab/git.py#L334

Reply:
Ah, I know what happened, reexamining this code. I had written and tested with the  `--reversed` flag in `jj log`, but accidentally broke this by not porting the flag usage over when I refactored to reuse `self.__cli_log`.

Will fix shortly!

Reply:
@maltejur: Can you confirm that e60d23ad fixes the issue for you? 🙏🏻